light/client.go

```diff
@@ -737,43 +731,69 @@ func (c *Client) bisection(
 			if depth == len(headerCache)-1 {
 				pivotHeight := trustedHeader.Height + (headerCache[depth].sh.Height-trustedHeader.
 					Height)*bisectionNumerator/bisectionDenominator
-				interimHeader, interimVals, err := c.fetchHeaderAndValsAtHeight(pivotHeight)
+				interimHeader, interimVals, err := c.signedHeaderAndValSetFrom(pivotHeight, source)
 				if err != nil {
 					return err
 				}
 				headerCache = append(headerCache, headerSet{interimHeader, interimVals})
 			}
 			depth++
 
-		case ErrInvalidHeader:
-			c.logger.Error("primary sent invalid header -> replacing", "err", err)
-			replaceErr := c.replacePrimaryProvider()
-			if replaceErr != nil {
-				c.logger.Error("Can't replace primary", "err", replaceErr)
-				// return original error
-				return fmt.Errorf("verify non adjacent from #%d to #%d failed: %w",
-					trustedHeader.Height, headerCache[depth].sh.Height, err)
-			}
-			newProviderHeader, newProviderVals, err := c.fetchHeaderAndValsAtHeight(newHeader.Height)
-			if err != nil {
-				return err
-			}
-			if !bytes.Equal(newProviderHeader.Hash(), newHeader.Hash()) || !bytes.Equal(newProviderVals.Hash(), newVals.Hash()) {
```

Comment:
I think this is wrong. Old primary sent us invalid header. By requiring that invalid header == newProviderHeader, we essentially checking that newProviderHeader is invalid too!

Reply:
Hmm, here is my thinking, VerifyHeader isn't necessarily used to get the latest header but to verify any header and val set which may have not come from the primary provider, so when a primary provider is being replaced I thought we should check that we're bisecting towards the same untrusted header else there is no point in continuing. You are right that if we use `VerifyHeaderAtHeight` that we take the header from this "faulty" provider in which case we should restart - it kind of depends if we want to get the header at a certain height, or verify an existing header

Reply:
In this implementation, if we were to do `VerifyHeaderAtHeight` it would return the header from the faulty provider and also an error:
```
err := fmt.Errorf("replacement provider has a different header: %X and/or vals: %X at height: %d"+
					"to the one being verified", newProviderHeader.Hash(), newProviderVals.Hash(), newHeader.Height)
```

which I think is okay - the user would just have to run it again with the new provider

Reply:
hmm, I'm on the side of removing though - too much complexity already

Reply:
if someone wanted to verify a provided header and we didn't have this case and completely restarted, the new provider could end up verifying a different header to the one initially requested to verify

Reply:
sure I understand the complexity part - so would the light client just return an error if it came across an invalid header and had to swap out the primary provider?

Reply:
I think in case where VerifyHeader(h) is called and h is invalid!, we should return the ErrHeaderInvalid immediately without trying to replace a provider

Reply:
+ we need the same check in sequence function. overwise, we don't know if we're verifying towards the same header

Reply:
sure but I mean the case where we run `VerifyHeader(h)` and h may be valid but during the process of verification, the primary returns an invalid header

Reply:
Okay, I reverted back the change. We can deal with this later in a separate PR.